### ISSUE TITLE
Airbrake.notify method is not safe in the rake tasks.

### ISSUE THREAD

**Issue by t-deetiw** [airbrake/airbrake#695]
For rake tasks, if there is any exception then airbrake gem will send the airbrake using notify_sync method.
But some where in the code flow while executing the rake task if Airbrake.notify is used and after which program gets terminated then airbrake is being missed.

eg.

```ruby
Class User
 # sample method to catch exception on some scenario
  def self.test
      begin
        raise 'Testing'
      rescue => e
         Airbrake.notify(e)
      end
  end
end

# Rake Task
namespace :cron do
  task test: :environment do
    User.test
  end
end
```

For normal web requests, it will be safe to use ``` User.test``` 
But for rake task its problematic.

**Comment by jeroenj** [comment]
We're also running against this probleem.

Adding a sleep below the code makes sure the notification gets sent to Airbrake.

The above example would be written like this:

```ruby
# Rake Task
namespace :cron do
  task test: :environment do
    User.test
    sleep 2
  end
end
```

**Comment by t-deetiw** [comment]
Yeah, Adding some sleep time will work. 
But adding it in each and every rake tasks would not be a clean practice.

Also, i think for rake tasks/runner scripts, we should make this gem flexible enough to accommodate below points.
1.  Always send airbrake synchronously since it will be running in the background in most of cases.
2.  Or We can make one more configurable option at application level to always send airbrake synchronously, in scope of rake task/ or runner script.

**Comment by jeroenj** [comment]
I think we're also having this problem in our resque workers (as the job process gets exited as soon as it finishes).

We're using the `sleep` call on most crucial parts now.

I'm not sure (yet) how the background jobs for Airbrake work but I think exiting the parent processes (the rake task, resque job, etc.) need to wait untill Airbrake has finished going through all the jobs in its queue.

**Comment by t-deetiw** [comment]
```
   # lib/airbrake/rake/task_ext.rb
   # execute method override for Rake::Task

    def execute(args = nil)
      execute_without_airbrake(args)
    rescue Exception => ex
      notify_airbrake(ex, args)
      raise ex
    end


    # I think, we can add some logic to clear the pending airbrake before rake task terminates inside ensure block.

    def execute(args = nil)
      execute_without_airbrake(args)
    rescue Exception => ex
      notify_airbrake(ex, args)
      raise ex
    ensure
       # some logic to clear the AsyncSender Queue, if there is any pending airbrake
    end
```

**Comment by kyrylo** [comment]
I believe this is due to buggy implementation of `Airbrake.notify_sync`, which became bugged when we introduced promises for `Airbrake.notfiy`. I am going to prepare a fix and I would be very thankful if any of you could help me and test it with your projects.

**Comment by kyrylo** [comment]
I'm sorry to misinform you. I confused this issue with another (undocumented) issue. Please disregard what I said earlier, it's absolutely irrelevant.

@t-deetiw
Now, to answer the outlined issue, if you know 100% that your script is going to exit after performing the task, you could close the notifier with [`Airbrake.close`](https://github.com/airbrake/airbrake-ruby#airbrakeclose). It doesn't have to be in an `at_exit` hook (as shown in the example there). Would that work for you?

@jeroenj 
Our Resque integration uses `Airbrake.notify_sync`, (which is not broken btw 😅) due to this exact reason. If you handle some code manually, I suggest to try using `Airbrake.notify_sync` when you handle your code manually. The worker is already running in a separate process, so it shouldn't be a big issue.

**Comment by t-deetiw** [comment]
@kyrylo 
Thanks a lot for the solution.

But only thing is, we have to be explicit everywhere to close the notifier to dispatch pending airbrakes, which we would prefer more if it's handled in the gem itself implicitly.
From my understanding it's always better to close the notifier ```at_exit``` hook.

**Comment by kyrylo** [comment]
I think this makes sense. Before making such a change, could you please verify whether this actually works?

**Comment by t-deetiw** [comment]
@kyrylo 
Verified it. It's working :-)

**Comment by jeroenj** [comment]
@kyrylo Using `Airbrake.close` works as expected. :)

**Comment by jeroenj** [comment]
Awesome! 👌

**Comment by jeroenj** [comment]
@kyrylo is a new version planned to be released soon?

**Comment by kyrylo** [comment]
Just released v6.0.0, which includes this change.

https://github.com/airbrake/airbrake/blob/master/CHANGELOG.md#v600-march-21-2017

**Comment by jeroenj** [comment]
Awesome. Thanks. 👌
